src/components/PostDetails.tsx

```diff
@@ -1,96 +1,43 @@
 import React, { useEffect, useState } from 'react';
 import { Loader } from './Loader';
-import { NewCommentForm } from './NewCommentForm';
-
-import * as commentsApi from '../api/comments';
+// import { NewCommentForm } from './NewCommentForm';
 
 import { Post } from '../types/Post';
-import { Comment, CommentData } from '../types/Comment';
+import { useAppDispatch, useAppSelector } from '../app/hooks';
+import { loadComments, remove, setError } from '../features/commentsSlice';
+import { deleteComment } from '../api/comments';
+import { NewCommentForm } from './NewCommentForm';
 
 type Props = {
   post: Post;
 };
 
 export const PostDetails: React.FC<Props> = ({ post }) => {
-  const [comments, setComments] = useState<Comment[]>([]);
-  const [loaded, setLoaded] = useState(false);
-  const [hasError, setError] = useState(false);
-  const [visible, setVisible] = useState(false);
-
-  function loadComments() {
-    setLoaded(false);
-    setError(false);
-    setVisible(false);
-
-    commentsApi.getPostComments(post.id)
-      .then(setComments) // save the loaded comments
-      .catch(() => setError(true)) // show an error when something went wrong
-      .finally(() => setLoaded(true)); // hide the spinner
-  }
-
-  useEffect(loadComments, [post.id]);
-
-  // The same useEffect with async/await
-  /*
-  async function loadComments() {
-    setLoaded(false);
-    setVisible(false);
-    setError(false);
+  const {
+    comments,
+    hasError,
+    loaded,
+  } = useAppSelector(state => state.comments);
 
-    try {
-      const commentsFromServer = await commentsApi.getPostComments(post.id);
-
-      setComments(commentsFromServer);
-    } catch (error) {
-      setError(true);
-    } finally {
-      setLoaded(true);
-    }
-  };
+  const [visible, setVisible] = useState(false);
+  const dispatch = useAppDispatch();
 
   useEffect(() => {
-    loadComments();
-  }, []);
+    dispatch(loadComments(post.id));
+    setVisible(false);
+  }, [post.id]);
 
-  useEffect(loadComments, [post.id]); // Wrong!
-  // effect can return only a function but not a Promise
-  */
+  const handleDeleteComment = (commentId: number) => {
+    dispatch(remove(commentId));
 
-  const addComment = async ({ name, email, body }: CommentData) => {
     try {
-      const newComment = await commentsApi.createComment({
-        name,
-        email,
-        body,
-        postId: post.id,
-      });
-
-      setComments(
-        currentComments => [...currentComments, newComment],
-      );
-
-      // setComments([...comments, newComment]);
-      // works wrong if we wrap `addComment` with `useCallback`
-      // because it takes the `comments` cached during the first render
-      // not the actual ones
-    } catch (error) {
-      // we show an error message in case of any error
-      setError(true);
+      deleteComment(commentId);
+    } catch {
+      dispatch(loadComments(post.id));
```

Comment:
why do you need `loadComments` if delete throws an error? maybe it would be better to create additional action fore deleting?

Reply:
Hi, we need to remove comment immediately from the page. To avoid waiting for the response I decided delete comment from the store and then send request to the server. But if the some error occures, we need to reload actual list of comments from the server and put it into store.